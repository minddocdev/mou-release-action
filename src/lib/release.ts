import * as fs from 'fs';
import * as path from 'path';

import * as core from '@actions/core';
import * as github from '@actions/github';

import { GitHubOctokit } from '../types';

export function renderReleaseName(releaseVersion: string, app?: string): string {
  return `${app ? `${app}@` : ''}${releaseVersion}`.trim();
}

function createReleaseNotes(
  releaseNotesFilepath: string,
  releaseNotesLanguageTags: string,
): string {
  return releaseNotesLanguageTags
    .split(',')
    .map((entry) => {
      const tag = entry.trim();
      const filepath = releaseNotesFilepath.replace('{LANGUAGE_TAG}', tag);
      const releaseNotes = fs.readFileSync(filepath, 'utf8');
      return `${tag}\n${releaseNotes}\n`;
    })
    .join('\n');
}

export function renderReleaseBody(
  templatePath: string,
  app: string,
  releaseVersion: string,
  changes = '',
  tasks = '',
  pullRequests = '',
  releaseNotesFilepath = '',
  releaseNotesLanguageTags = '',
): string {
  const { repo } = github.context.repo;
  let body = fs
    .readFileSync(path.resolve('/home/runner/work', repo, repo, '.github', templatePath), 'utf8')
    .replace(/\$APP/g, app)
    .replace(/\$VERSION/g, releaseVersion);

  body = body.replace(/\$CHANGES/g, changes);
  body = body.replace(/\$TASKS/g, tasks);
  body = body.replace(/\$PULL_REQUESTS/g, pullRequests);

  if (releaseNotesFilepath !== '') {
    body = body.replace(
      /\$RELEASE_NOTES/g,
      createReleaseNotes(releaseNotesFilepath, releaseNotesLanguageTags),
    );
  }

  return body;
}

export async function createGitTag(
  octokit: InstanceType<typeof GitHubOctokit>,
  tag: string,
): Promise<void> {
  const { owner, repo } = github.context.repo;
  const { sha } = github.context;

  const response = await octokit.git.createRef({
    owner,
    repo,
    sha,
    ref: `refs/tags/${tag}`,
  });
  if (response.status !== 201) {
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    throw new Error(`Unable to create tag ${tag}. Github returned status: ${response.status}`);
  }
}

export async function createGithubRelease(
  octokit: InstanceType<typeof GitHubOctokit>,
  tag: string,
  name: string,
  body: string,
  draft: boolean,
  prerelease: boolean,
): Promise<void> {
  const { owner, repo } = github.context.repo;
  // Create a release
  // API Documentation: https://developer.github.com/v3/repos/releases/#create-a-release
  // Octokit Documentation: https://octokit.github.io/rest.js/#octokit-routes-repos-create-release
  const createReleaseResponse = await octokit.repos.createRelease({
    owner,
    repo,
    tag_name: tag,
    name,
    body,
    draft,
    prerelease,
  });

  // Get the ID, html_url, and upload URL for the created Release from the response
  const {
    data: { id: releaseId, html_url: htmlUrl, upload_url: uploadUrl },
  } = createReleaseResponse;

  core.setOutput('release_id', releaseId.toString());
  core.setOutput('html_url', htmlUrl);
  core.setOutput('upload_url', uploadUrl);
}
