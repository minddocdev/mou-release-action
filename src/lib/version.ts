import semver from 'semver';

import * as core from '@actions/core';
import * as github from '@actions/github';

import { GitHubOctokit } from '../types';
import { Octokit } from '@octokit/rest';

// See semver.ReleaseType
export enum VersionType {
  major = 'major',
  minor = 'minor',
  patch = 'patch',
  prerelease = 'prerelease',
}

const findReleaseTag = async (
  octokit: InstanceType<typeof Octokit>,
  // eslint-disable-next-line @typescript-eslint/ban-types
  matchFunction: (release: Record<string, unknown>) => {},
) => {
  const { owner, repo } = github.context.repo;

  // Using pagination: https://octokit.github.io/rest.js/v17#pagination
  const listReleasesOptions = octokit.repos.listReleases.endpoint.merge({
    owner,
    repo,
  });

  // Look for the earliest release that matches the given condition
  /* eslint-disable no-restricted-syntax */
  for await (const response of octokit.paginate.iterator<any>(listReleasesOptions)) {
    for (const release of response.data) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      if (matchFunction(release)) return release.tag_name as string;
    }
  }
  /* eslint-enable no-restricted-syntax */
  return undefined;
};

export async function bumpVersion(
  octokit: InstanceType<typeof GitHubOctokit>,
  tagPrefix: string,
  nextVersionType = VersionType.patch,
  publishedTag?: string,
  bumpProtection = false,
): Promise<string> {
  const publishedVersion = publishedTag ? publishedTag.replace(tagPrefix, '') : undefined;
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
  const matchesTagPrefix = (release: any) => release.tag_name.startsWith(tagPrefix);
  // Load latest production tag from published releases
  const lastTag = await findReleaseTag(octokit, matchesTagPrefix);
  const lastVersion = lastTag ? lastTag.replace(tagPrefix, '') : '0.0.0';

  let releaseType = nextVersionType;
  let newVersion: string | null = null;
  if (releaseType === VersionType.prerelease) {
    // Bump release candidate as 'prerelease' if detected as next release type
    newVersion = semver.inc(lastVersion, releaseType, 'rc');
  } else {
    // 'major', 'minor' or 'patch' needs to be bumped
    // Override to 'patch' if bump protection is triggered
    if (
      bumpProtection &&
      publishedVersion &&
      // MINOR protection (if there was already a previous MINOR bump)
      ((nextVersionType === VersionType.minor &&
        !semver.satisfies(
          lastVersion,
          // ^1.2	is >=1.2.0 <2.0.0
          `^${semver.major(publishedVersion)}.${semver.minor(publishedVersion)}`,
        )) ||
        // MAJOR protection (if there was already a previous MAJOR bump)
        (nextVersionType === VersionType.major &&
          !semver.satisfies(
            lastVersion,
            // ^1	is >=1.0.0 <2.0.0
            `^${semver.major(publishedVersion)}`,
          )))
    ) {
      // Detect if there was a minor or major update between the latest production tag (baseTag)
      // and the latest tag (the one that will be bumped), if bump is not a patch.
      // If a major or minor update already happened, perform a patch instead.
      // Production deployments will never bump a minor or major more than once, while internal
      // tags can be bumped
      releaseType = VersionType.patch;
    }
    newVersion = semver.inc(lastVersion, releaseType);
  }

  if (newVersion === null) {
    throw new Error(`Unable to perform a ${releaseType} bump to version ${lastVersion}`);
  }
  const newTag = `${tagPrefix}${newVersion}`;

  core.setOutput('previous_tag', lastTag || '');
  core.setOutput('previous_version', lastVersion);
  core.setOutput('new_tag', newTag);
  core.setOutput('new_version', newVersion);
  core.setOutput('release_type', releaseType);
  return newTag;
}

export async function retrieveLastReleasedVersion(
  octokit: any,
  tagPrefix: string,
): Promise<string | undefined> {
  const isVersionReleased = (release) => {
    const { prerelease, draft, tag_name: tagName } = release;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    return !draft && !prerelease && tagName.startsWith(tagPrefix);
  };
  const lastPublishedTag = await findReleaseTag(octokit, isVersionReleased);
  core.setOutput('base_tag', lastPublishedTag || '');
  return lastPublishedTag;
}
