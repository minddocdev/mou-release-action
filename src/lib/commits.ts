import * as core from '@actions/core';
import * as github from '@actions/github';

import { GitHubOctokit } from '../types';
import { VersionType } from './version';

// Octokit commit type subset
interface Commit {
  username: string;
  userUrl: string;
  commitUrl: string;
  message: string;
  sha: string;
}

/**
 * Generate a Markdown string based on commit information.
 * @param commits - The JSON array with Github commit info.
 */
export async function commitParser(
  octokit: InstanceType<typeof GitHubOctokit>,
  baseRef = 'master',
  taskPrefix = 'JIRA-',
  taskBaseUrl?: string,
  commitScope?: string,
): Promise<{
  nextVersionType: VersionType;
  changes: string;
  tasks: string;
  pullRequests: string;
}> {
  const commitGroups: {
    [index: string]: { title: string; commits: Commit[] };
  } = {
    // A new feature
    feat: {
      title: '**:zap: Features**',
      commits: [],
    },
    // A bug fix
    fix: {
      title: '**:wrench: Fixes**',
      commits: [],
    },
    // A code change that improves performance
    perf: {
      title: '**:runner: Performance**',
      commits: [],
    },
    // Documentation only changes
    docs: {
      title: '**:books: Documentation**',
      commits: [],
    },
    // Changes that do not affect the meaning of the code (lint changes)
    style: {
      title: '**:nail_care: Style**',
      commits: [],
    },
    // A code change that neither fixes a bug nor adds a feature
    refactor: {
      title: '**:mountain: Refactors**',
      commits: [],
    },
    // Adding missing tests or correcting existing tests
    test: {
      title: '**:traffic_light: Tests**',
      commits: [],
    },
    // Changes that affect the build system or external development dependencies
    chore: {
      title: '**:construction: Maintenance**',
      commits: [],
    },
    // As an alternative to 'chore', but with very similar meaning
    build: {
      title: '**:construction_worker: Build**',
      commits: [],
    },
    // Changes for CI configuration files and scripts (e.g. Github CI, helm values...)
    ci: {
      title: '**:runner: CI**',
      commits: [],
    },
  };
  const uncategorizedCommits: Commit[] = [];

  const changes: string[] = [];
  const tasks = new Set();
  const pullRequests: string[] = [];
  let nextVersionType = VersionType.patch;

  const { owner, repo } = github.context.repo;
  const compareCommitsResponse = await octokit.repos.compareCommits({
    owner,
    repo,
    base: baseRef,
    head: github.context.sha,
  });
  const {
    data: { commits },
  } = compareCommitsResponse;

  const categorizeCommit = (commit: Commit) => {
    const { message } = commit;
    // Skip if scope check is required and commit does not have it
    if (commitScope && !message.includes(`(${commitScope}):`)) {
      return;
    }
    // Check if commit message matches to any of the defined categories
    let categoryMatch = false;
    Object.keys(commitGroups).some((category) => {
      // Match with or without scope
      if (message.startsWith(`${category}:`) || message.startsWith(`${category}(`)) {
        commitGroups[category].commits.push(commit);
        categoryMatch = true;
        return true;
      }
      return false;
    });
    if (!categoryMatch) uncategorizedCommits.push(commit);
  };

  const prRegExp = new RegExp('(\\(#\\d+\\))', 'gmi');
  const taskRegExp = new RegExp(`${taskPrefix}\\d+`, 'gmi');
  const majorRegExp = new RegExp(`(#MAJOR$)`, 'gmi');
  commits.forEach((githubCommit) => {
    const username = githubCommit.author?.login ?? 'minddocbot';
    const userUrl = githubCommit.author?.html_url ?? 'https://github.com/minddocbot';
    const commitUrl = githubCommit.html_url;
    const { message } = githubCommit.commit;
    const { sha } = githubCommit;
    const commit: Commit = { username, userUrl, commitUrl, message, sha };

    // Retrieve PR link information
    const prMatch = prRegExp.exec(message);
    if (prMatch) prMatch.slice(1).forEach((pr) => pullRequests.push(pr.replace(/(\(|\)|#)/g, '')));
    // Retrieve task information
    // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
    const taskMatch = message.match(taskRegExp);
    if (taskMatch) taskMatch.forEach((task) => tasks.add(task));
    // Retrieve specific bump key words
    const majorMatch = majorRegExp.exec(message);
    if (majorMatch) nextVersionType = VersionType.major;

    // Detect if commit is a Github squash. In that case, convert body
    // in multiple single line commits and parse
    if (/\* .*\n/.test(message)) {
      const messageLines = message.split('* ');
      // Categorize all commits except first one
      messageLines.forEach((messageLine) =>
        categorizeCommit({
          username,
          userUrl,
          commitUrl,
          sha,
          message: messageLine.trim(),
        }),
      );
    } else {
      categorizeCommit(commit);
    }
  });

  let changesMd = '';

  const formatCommit = (commit: Commit) => {
    const { username, userUrl, sha, commitUrl } = commit;
    let { message } = commit;
    // Only take into account the commit title
    [message] = message.split('\n');
    // Detect if commit message has Angular format
    if (/(\w+\([a-zA-Z_-]+\)|\w+|\([a-zA-Z_-]+\)):/.test(message)) {
      // Remove group information for changelog (e.g. messages with categories)
      message = message.split(':')[1].trim();
    }
    // Always capitalize commit messages
    message = `${message[0].toUpperCase()}${message.slice(1)}`;
    // Add to global change markdown
    changesMd = `${changesMd}- ${message} - [${sha.substring(
      0,
      8,
    )}](${commitUrl})([@${username}](${userUrl}))\n`;
    // Add to global commit sha list
    changes.push(sha);
  };

  uncategorizedCommits.forEach(formatCommit);

  Object.keys(commitGroups).forEach((category) => {
    const { title, commits: groupCommits } = commitGroups[category];
    if (groupCommits.length !== 0) {
      changesMd = `${changesMd}\n${title}\n`;
      groupCommits.forEach(formatCommit);
    }
  });

  core.setOutput('changes', JSON.stringify(changes));
  core.setOutput('tasks', JSON.stringify([...tasks]));
  core.setOutput('pull_requests', JSON.stringify(pullRequests));

  // Set bump type to minor if there is at least one 'feat' commit
  if (nextVersionType === VersionType.patch && commitGroups.feat.commits.length > 0) {
    nextVersionType = VersionType.minor;
  }
  core.setOutput('change_type', nextVersionType);

  return {
    nextVersionType,
    changes: changesMd.trim(),
    tasks: [...tasks]
      .map((task) => task as string)
      .map((task) => `[${task}](${taskBaseUrl || `https://${owner}.atlassian.net/browse`}/${task})`)
      .join(', '),
    pullRequests: pullRequests
      .map((pr) => `[#${pr}](https://github.com/${owner}/${repo}/pull/${pr})`)
      .join(', '),
  };
}
