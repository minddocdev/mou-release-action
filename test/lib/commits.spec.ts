import { setOutput } from '@actions/core';
import { commitParser } from '@minddocdev/mou-release-action/lib/commits';
import { VersionType } from '@minddocdev/mou-release-action/lib/version';

jest.mock('@actions/github', () => ({
  context: {
    repo: {
      owner: 'theowner',
      repo: 'therepo',
    },
  },
}));
jest.mock('@actions/core');

const author = {
  login: 'darioblanco',
  html_url: 'https://authorurl',
};

// eslint-disable-next-line @typescript-eslint/naming-convention
const html_url = 'https://commiturl';
const sha = '62ec8ea713fdf14e4abaef3d7d5138194dec49ce';

describe('commit', () => {
  const compareCommitsResponse = {
    data: {
      commits: [
        {
          author,
          html_url,
          sha,
          commit: { message: 'feat:super feature' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'fix: My fix' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'perf: set additional performance steps' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'docs(my-app): document everything' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'style: awesome style' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'refactor(my-app): one does not simply refactor' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'test: Tests are good' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'chore: somebody has to keep things going' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'build: like chore but fancier' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'ci: ok this is a CI change' },
        },
        {
          author,
          html_url,
          sha,
          commit: { message: 'uncategorized commit' },
        },
      ],
    },
  };
  const compareCommits = jest.fn(() => compareCommitsResponse);

  beforeEach(() => compareCommits.mockClear());

  test('render commits diff for each category', async () => {
    const github = { repos: { compareCommits } };
    const { changes, nextVersionType, tasks, pullRequests } = await commitParser(github as any);
    expect(setOutput).toBeCalledWith(
      'changes',
      JSON.stringify(compareCommitsResponse.data.commits.map((commit) => commit.sha)), // 8 commits
    );
    expect(setOutput).toBeCalledWith('tasks', '[]');
    expect(setOutput).toBeCalledWith('pull_requests', '[]');
    expect(changes).toMatchSnapshot();
    expect(nextVersionType).toBe(VersionType.minor);
    expect(tasks).toBe('');
    expect(pullRequests).toBe('');
  });

  test('render commits diff for each category using the default author', async () => {
    const compareCommitsCategoryResponse = {
      data: {
        commits: [
          {
            author: undefined,
            html_url,
            sha,
            commit: {
              message: 'commit message',
            },
          },
        ],
      },
    };
    const github = { repos: { compareCommits: jest.fn(() => compareCommitsCategoryResponse) } };
    const { changes, nextVersionType, tasks, pullRequests } = await commitParser(github as any);
    expect(setOutput).toBeCalledWith(
      'changes',
      JSON.stringify(compareCommitsCategoryResponse.data.commits.map((commit) => commit.sha)), // 8 commits
    );
    expect(setOutput).toBeCalledWith('tasks', '[]');
    expect(setOutput).toBeCalledWith('pull_requests', '[]');
    expect(changes).toMatchSnapshot();
    expect(nextVersionType).toBe(VersionType.patch);
    expect(tasks).toBe('');
    expect(pullRequests).toBe('');
  });

  test('render commits diff when scope is required', async () => {
    const github = { repos: { compareCommits } };
    const { changes, nextVersionType, tasks, pullRequests } = await commitParser(
      github as any,
      'v1.0.0',
      'JIRA-',
      undefined,
      'my-app',
    );
    expect(setOutput).toBeCalledWith(
      'changes',
      '["62ec8ea713fdf14e4abaef3d7d5138194dec49ce","62ec8ea713fdf14e4abaef3d7d5138194dec49ce"]',
    );
    expect(setOutput).toBeCalledWith('tasks', '[]');
    expect(setOutput).toBeCalledWith('pull_requests', '[]');
    expect(changes).toMatchSnapshot();
    expect(nextVersionType).toBe(VersionType.patch);
    expect(tasks).toBe('');
    expect(pullRequests).toBe('');
  });

  [undefined, 'http://my-task-url'].forEach((taskBaseUrl) =>
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    test(`render gh squashed commits with scope, PRs and tasks for ${taskBaseUrl!}`, async () => {
      const commitMessage =
        'feat(auth): main commit of my PR (#1716)\n\n' +
        '* feat(auth): set login endpoint controller\n\n' +
        '* test(auth): [JIRA-123] add integration test for login endpoint #MAJOR\n\n' +
        '* fix(auth): [JIRA-123] set secure and http only options\n\n' +
        '* perf(auth): add additional fake performance\n\n' +
        'This is the body of the previous commit\n\n' +
        'And this is the footer\n\n' +
        '* (auth): Set values for staging and production\n\n' +
        '* Address comment from PR\n\n' +
        'This is the body of the previous commit\n\n' +
        '* feat(auth): set expiration of the cookie to the amount of time of the token\n\n' +
        '* fix(auth): remove joi validation since it does not accept localhost\n\n' +
        'Closes: JIRA-2772,JIRA-2773';
      const compareSquashedCommitsResponse = {
        data: {
          commits: [
            {
              author,
              html_url,
              sha,
              commit: {
                message: commitMessage,
              },
            },
          ],
        },
      };
      const github = { repos: { compareCommits: jest.fn(() => compareSquashedCommitsResponse) } };
      const { changes, nextVersionType, tasks, pullRequests } = await commitParser(
        github as any,
        'v1.0.0',
        'JIRA-',
        taskBaseUrl,
        'auth',
      );
      expect(setOutput).toBeCalledWith(
        'changes',
        JSON.stringify(
          commitMessage
            .split('* ')
            .slice(1)
            .map(() => sha),
        ), // 6 commits
      );
      expect(setOutput).toBeCalledWith('tasks', '["JIRA-123","JIRA-2772","JIRA-2773"]');
      expect(setOutput).toBeCalledWith('pull_requests', '["1716"]');
      expect(changes).toMatchSnapshot();
      expect(nextVersionType).toBe(VersionType.major);
      const jiraBaseUrl = taskBaseUrl || 'https://theowner.atlassian.net/browse';
      expect(tasks).toBe(
        `[JIRA-123](${jiraBaseUrl}/JIRA-123), [JIRA-2772](${jiraBaseUrl}/JIRA-2772), [JIRA-2773](${jiraBaseUrl}/JIRA-2773)`,
      );
      expect(pullRequests).toBe('[#1716](https://github.com/theowner/therepo/pull/1716)');
    }),
  );

  test('render github squashed commits without scope, PRs and tasks', async () => {
    const commitMessage =
      'Title of my PR\n\n' +
      '* set login endpoint controller\n\n' +
      '* add integration test for login endpoint\n\n';
    const compareSquashedCommitsResponse = {
      data: {
        commits: [
          {
            author,
            html_url,
            sha,
            commit: {
              message: commitMessage,
            },
          },
        ],
      },
    };
    const github = { repos: { compareCommits: jest.fn(() => compareSquashedCommitsResponse) } };
    const { changes, nextVersionType, tasks, pullRequests } = await commitParser(
      github as any,
      'v1.0.0',
      'JIRA-',
    );
    expect(setOutput).toBeCalledWith(
      'changes',
      '["62ec8ea713fdf14e4abaef3d7d5138194dec49ce",' +
        '"62ec8ea713fdf14e4abaef3d7d5138194dec49ce",' +
        '"62ec8ea713fdf14e4abaef3d7d5138194dec49ce"]',
    );
    expect(setOutput).toBeCalledWith('tasks', '[]');
    expect(setOutput).toBeCalledWith('pull_requests', '[]');
    expect(changes).toMatchSnapshot();
    expect(nextVersionType).toBe(VersionType.patch);
    expect(tasks).toBe('');
    expect(pullRequests).toBe('');
  });
});
