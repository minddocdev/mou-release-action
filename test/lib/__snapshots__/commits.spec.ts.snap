// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit render commits diff for each category 1`] = `
"- Uncategorized commit - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:zap: Features**
- Super feature - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:wrench: Fixes**
- My fix - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:runner: Performance**
- Set additional performance steps - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:books: Documentation**
- Document everything - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:nail_care: Style**
- Awesome style - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:mountain: Refactors**
- One does not simply refactor - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:traffic_light: Tests**
- Tests are good - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:construction: Maintenance**
- Somebody has to keep things going - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:construction_worker: Build**
- Like chore but fancier - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:runner: CI**
- Ok this is a CI change - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))"
`;

exports[`commit render commits diff for each category using the default author 1`] = `"- Commit message - [62ec8ea7](https://commiturl)([@minddocbot](https://github.com/minddocbot))"`;

exports[`commit render commits diff when scope is required 1`] = `
"**:books: Documentation**
- Document everything - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:mountain: Refactors**
- One does not simply refactor - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))"
`;

exports[`commit render gh squashed commits with scope, PRs and tasks for http://my-task-url 1`] = `
"- Set values for staging and production - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:zap: Features**
- Main commit of my PR (#1716) - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Set login endpoint controller - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Set expiration of the cookie to the amount of time of the token - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:wrench: Fixes**
- [JIRA-123] set secure and http only options - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Remove joi validation since it does not accept localhost - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:runner: Performance**
- Add additional fake performance - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:traffic_light: Tests**
- [JIRA-123] add integration test for login endpoint #MAJOR - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))"
`;

exports[`commit render gh squashed commits with scope, PRs and tasks for undefined 1`] = `
"- Set values for staging and production - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:zap: Features**
- Main commit of my PR (#1716) - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Set login endpoint controller - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Set expiration of the cookie to the amount of time of the token - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:wrench: Fixes**
- [JIRA-123] set secure and http only options - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Remove joi validation since it does not accept localhost - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:runner: Performance**
- Add additional fake performance - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:traffic_light: Tests**
- [JIRA-123] add integration test for login endpoint #MAJOR - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))"
`;

exports[`commit render github squashed commits without scope, PRs and tasks 1`] = `
"- Title of my PR - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Set login endpoint controller - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))
- Add integration test for login endpoint - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))"
`;
